### PR TITLE
feat(tui): recognize keypad keys when using kitty keyboard protocol

### LINKED ISSUE #19174: Broken Keypad Enter
### Neovim version (nvim -v)

0.7.2

### Vim (not Nvim) behaves the same?

no, vim-9.0.0000-1

### Operating system/version

Arch Linux x86-64

### Terminal name/version

kitty 0.25.2 

### $TERM environment variable

xterm-kitty

### Installation

pacman

### How to reproduce the issue

Press keypad (numpad) enter in normal or insert mode with numlock off.

### Expected behavior

Enter keystroke

### Actual behavior

An  (æ) character being printed

### PR DESCRIPTION
**Pull request by zeertzjq**
Fix #19174

Based on [`8be0322` (#16570)](https://github.com/neovim/neovim/pull/16570/commits/8be03229c01bd85a93a071ed379999cda53dd96c)

### CODE REVIEW

File: src/nvim/tui/input_defs.h [hunk 1]
```diff
@@ -0,0 +1,92 @@
+#ifndef NVIM_TUI_INPUT_DEFS_H
+#define NVIM_TUI_INPUT_DEFS_H
+
+typedef enum {
+  KITTY_KEY_CAPS_LOCK            = 57358,
```
Comment (justinmk) [inline src/nvim/tui/input_defs.h:5]: Not excited about adding lots of special case code for kitty. Are we going to be stuck with this forever or is it possible that libvterm would support it?

Comment (zeertzjq) [reply src/nvim/tui/input_defs.h:5]: Do you mean libtermkey?

Comment (justinmk) [reply src/nvim/tui/input_defs.h:5]: I meant `libtickit` actually :) https://www.leonerd.org.uk/code/libtermkey/

**Comment by justinmk** [comment]
Trying to get an idea of what we lose if we don't add this. Does kitty support CSIu / modifyotherkeys / anything like that? Or it only supports its own custom protocol? How many terminals have kitty protocol support?

**Comment by zeertzjq** [comment]
Kitty keyboard protocol seems to be the only protocol that sends keypad keys differently. Other protocols just send keypad number as numbers, keypad arrows as the same sequence as normal arrow keys, etc..

According to <https://sw.kovidgoyal.net/kitty/keyboard-protocol/>, kitty keyboard protocol is also supported by foot and WezTerm.

Without this PR, keypads keys do not work in terminal emulators that support kitty keyboard protocol.

**Comment by kovidgoyal** [comment]
libtickit/leonerd's CSIu is full of bugs, dont use it. Read https://sw.kovidgoyal.net/kitty/keyboard-protocol/#bugs-in-fixterms

It takes literally one lookup table https://sw.kovidgoyal.net/kitty/keyboard-protocol/#functional-key-definitions and 35 lines of code https://github.com/kovidgoyal/kitty/blob/master/kitty/key_encoding.py#L310 to decode keys from the kitty keyboard protocol. You dont need an external library, let alone a broken one for this.

File: src/nvim/tui/input.c [hunk 2]
```diff
@@ -206,19 +271,46 @@ static void tinput_enqueue(TermInput *input, char *buf, size_t size)
...
   char *ptr = key->utf8;
 
-  while (*ptr) {
-    if (*ptr == '<') {
-      len += (size_t)snprintf(buf + len, sizeof(buf) - len, "<lt>");
-    } else {
-      buf[len++] = *ptr;
+  if (key->code.codepoint >= 0xE000 && key->code.codepoint <= 0xF8FF
+      && map_has(KittyKey, cstr_t)(&kitty_key_map, (KittyKey)key->code.codepoint)) {
+    handle_kitty_key_protocol(input, key);
```
Comment (justinmk) [inline src/nvim/tui/input.c:303]: So this works automatically based on the codepoint? Neither users nor Nvim need to set a terminal option? Would this be potentially usable by any terminal emulator that allows configuring arbitrary "key => sequenc" mappings?

Do we need a doc near `:help tui-input`, similar to `:help tui-csiu` ?

**Review by justinmk — APPROVED** [review]

**Comment by justinmk** [comment]
does this close #14400 ? 

cc @erw7

**Comment by zeertzjq** [comment]
This still does not recognize <kbd>Super</kbd> modifier.

**Comment by erw7** [comment]
If only `SHIFT`, `ALT`, and `CTRL` were supported as modifier keys, then the `handle_kitty_key_protocol()` would have been unnecessary (libtermkey doesn't handle it?).

By the way, is this really necessary? If users really need this, why don't they just set up the following on their own?

```vim
noremap <Char-57364> <F1>
noremap <Char-57365> <F2>
...
```

**Comment by zeertzjq** [comment]
A downside of using mapping is that they don't work a hit-enter prompt.

**Comment by justinmk** [comment]
> > By the way, is this really necessary? If users really need this, why don't they just set up the following on their own?
> > ```viml
> > noremap <Char-57364> <F1>
> > noremap <Char-57365> <F2>
> > ...
> > ```
> 
> A downside of using mappings is that they don't work at places where mappings are not allowed, e.g. at hit-enter prompt or after pressing `z` in Normal mode or `getchar()`.

These are both good insights. Let's mention the above at `:help tui-input` or `:help tui-kitty`.

**Comment by zeertzjq** [comment]
I thought about this for hours and still cannot decide how to write this properly in docs.

**Comment by tobealive** [comment]
Hmm, it seems that it made it worse 🤔

For traveling my potato laptop is just the right tool as it can sometimes take a hit. But it sends kp_enter with it's regular enter key, so this issue was recognizable right away. 

- It went from **no** issues with <kbd>Enter</kbd> in `0.7.0`
- To: having to map `""` to `"<CR>"` in `0.7.2`
- To: a not working <kbd>Enter</kbd> key in `nightly`  (although it sends a correct `^M`?. But is not mapable - At least I couldn't figure it out.)
But using its equivalent <kbd>Ctrl+M</kbd> works, and is just fine. Still an issue we can't deny!

One min from the 🍠 that show all version. _(little info: enter which always sends **KP_enter** on this machine, **is a blank in screenkey**)_



https://user-images.githubusercontent.com/34311583/180609715-35379746-2d14-42d1-b80e-0b2e458c2a50.mp4



At first, I thought it's an issue with kitty, so I tried to map all kind of enter key variations inside my kitty conf. But it didn't work out, and the issue is also present in other tuis like alacrity or wezterm. No issues with tilix. I hope some of those info might help.

**Comment by zeertzjq** [comment]
Does mapping `<kEnter>` work?

**Comment by tobealive** [comment]
Reading this, not knowing of this mapping I already thought that it probably will do the job.
And it does. So simple. Somehow it didn't come to my mind to check for this.

Thanks for letting me know something new.

**Comment by tobealive** [comment]
Okay maybe not that simple. 

It's not working anymore after booting back into this mint installation. (I checked if I can get it working properly with an arch installation on this machine, which had more extensive errors regarding the `kp_enter`).  

This it what it screems when I try to <kbd>Enter</kbd> in a telescope dialog like in the video above>
```vim
  .../nightly/nvim-linux64/share/nvim/runtime/lua/vim/lsp.lua:1397: in function <.../nightly/nvim-linu
x64/share/nvim/runtime/lua/vim/lsp.lua:1389>
Error executing lua callback: ...tly/nvim-linux64/share/nvim/runtime/lua/vim/lsp/sync.lua:177: attempt to ge
t length of local 'prev_line' (a nil value)
stack traceback:
        ...tly/nvim-linux64/share/nvim/runtime/lua/vim/lsp/sync.lua:177: in function 'compute_start_range'
        ...tly/nvim-linux64/share/nvim/runtime/lua/vim/lsp/sync.lua:408: in function 'compute_diff'
        .../nightly/nvim-linux64/share/nvim/runtime/lua/vim/lsp.lua:457: in function 'incremental_changes'
        .../nightly/nvim-linux64/share/nvim/runtime/lua/vim/lsp.lua:500: in function 'fn'
        .../nightly/nvim-linux64/share/nvim/runtime/lua/vim/lsp.lua:177: in function 'for_each_buffer_client
'
        .../nightly/nvim-linux64/share/nvim/runtime/lua/vim/lsp.lua:1397: in function <.../nightly/nvim-lin
```

Unfortunately I do not know where to hook on to get proper debug information.

**Comment by page-down** [comment]
@zeertzjq
I noticed that `<S-Space>` doesn't work under kitty, while it does under iTerm2 (I think it's `XTMODKEYS`).

```vim
nnoremap <Space> <C-d>
nnoremap <S-Space> <C-u>
```

Is it possible to fix it this way as well? Thanks.
Tested with the latest neovim nightly.

https://github.com/kovidgoyal/kitty/discussions/5342#discussioncomment-3294884
> shift+space is `CSI 32 ; 2 ; 32 u`
> space without shift is `CSI 32 ; ; 32 u`

I would like to ask here first instead of opening a new issue.

**Comment by zeertzjq** [comment]
Nvim only enables "Disambiguate escape codes" progressive enhancement, and kitty doesn't seems to send <kbd>Shift</kbd>+<kbd>Space</kbd> differently in this mode.

**Pull request merged**